File: services/hierarchy/src/main/java/com/casemap/hierarchy/featurekey/FeatureKey.java
package com.casemap.hierarchy.featurekey;

import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

/**
 * featureKey 规则：领域/系统/场景/功能点
 */
public final class FeatureKey {

    public static final String SEPARATOR = "/";

    private final String domain;
    private final String system;
    private final String scene;
    private final String feature;

    private FeatureKey(String domain, String system, String scene, String feature) {
        this.domain = domain;
        this.system = system;
        this.scene = scene;
        this.feature = feature;
    }

    public String domain() {
        return domain;
    }

    public String system() {
        return system;
    }

    public String scene() {
        return scene;
    }

    public String feature() {
        return feature;
    }

    public List<String> asList() {
        return List.of(domain, system, scene, feature);
    }

    public String toKey() {
        return String.join(SEPARATOR, domain, system, scene, feature);
    }

    public static FeatureKey of(String domain, String system, String scene, String feature) {
        return new FeatureKey(
                normalize(domain, "领域"),
                normalize(system, "系统"),
                normalize(scene, "场景"),
                normalize(feature, "功能点")
        );
    }

    public static FeatureKey parse(String featureKey) {
        String raw = featureKey == null ? "" : featureKey.trim();
        if (raw.isEmpty()) {
            throw new FeatureKeyException("featureKey 不能为空");
        }
        String[] segments = raw.split(SEPARATOR, -1);
        if (segments.length != 4) {
            throw new FeatureKeyException(
                    "featureKey 必须为「领域/系统/场景/功能点」四级，当前 "
                            + segments.length + " 段：" + featureKey);
        }
        return of(segments[0], segments[1], segments[2], segments[3]);
    }

    public static String join(String domain, String system, String scene, String feature) {
        return of(domain, system, scene, feature).toKey();
    }

    public static String encodeForUrl(String featureKey) {
        return URLEncoder.encode(parse(featureKey).toKey(), StandardCharsets.UTF_8);
    }

    public static String decodeFromUrl(String encoded) {
        String decoded = URLDecoder.decode(Objects.requireNonNullElse(encoded, ""), StandardCharsets.UTF_8);
        return parse(decoded).toKey();
    }

    /**
     * 解析路径变量中的 featureKey。
     * 兼容未编码的四级路径、整段百分号编码，以及 Spring {@code {*featureKey}} 可能带上的前导斜杠。
     */
    public static String fromPathVariable(String featureKey) {
        String raw = featureKey == null ? "" : featureKey.trim();
        if (raw.startsWith(SEPARATOR)) {
            raw = raw.substring(1);
        }
        if (raw.contains("%")) {
            return decodeFromUrl(raw);
        }
        return parse(raw).toKey();
    }

    public static List<String> exampleQuoteKeys() {
        return Arrays.asList(
                "家装/报价/金额计算与汇总/数量价汇总",
                "家装/报价/金额计算与汇总/固定价汇总",
                "家装/报价/造价提交与审核/自动审核判定"
        );
    }

    public static boolean isValid(String featureKey) {
        try {
            parse(featureKey);
            return true;
        } catch (FeatureKeyException ex) {
            return false;
        }
    }

    private static String normalize(String segment, String levelLabel) {
        String value = segment == null ? "" : segment.trim();
        if (value.isEmpty()) {
            throw new FeatureKeyException(levelLabel + "不能为空");
        }
        if (value.contains(SEPARATOR)) {
            throw new FeatureKeyException(levelLabel + "不能包含分隔符「/」，原始值：" + segment);
        }
        if (value.contains("\n") || value.contains("\r")) {
            throw new FeatureKeyException(levelLabel + "不能包含换行");
        }
        return value;
    }
}

File: services/casemap/app/feature_key.py
from __future__ import annotations

from dataclasses import dataclass
from urllib.parse import quote_plus, unquote_plus

from app.errors import FeatureKeyError

SEPARATOR = "/"
EXAMPLE_QUOTE_KEYS = [
    "家装/报价/金额计算与汇总/数量价汇总",
    "家装/报价/金额计算与汇总/固定价汇总",
    "家装/报价/造价提交与审核/自动审核判定",
]


@dataclass(frozen=True)
class FeatureKey:
    domain: str
    system: str
    scene: str
    feature: str

    def as_list(self) -> list[str]:
        return [self.domain, self.system, self.scene, self.feature]

    def to_key(self) -> str:
        return SEPARATOR.join(self.as_list())


def _normalize(segment: str | None, level_label: str) -> str:
    value = "" if segment is None else segment.strip()
    if not value:
        raise FeatureKeyError(level_label + "不能为空")
    if SEPARATOR in value:
        raise FeatureKeyError(level_label + "不能包含分隔符「/」，原始值：" + str(segment))
    if "\n" in value or "\r" in value:
        raise FeatureKeyError(level_label + "不能包含换行")
    return value


def of(domain: str, system: str, scene: str, feature: str) -> FeatureKey:
    return FeatureKey(
        _normalize(domain, "领域"),
        _normalize(system, "系统"),
        _normalize(scene, "场景"),
        _normalize(feature, "功能点"),
    )


def parse(feature_key: str | None) -> FeatureKey:
    raw = "" if feature_key is None else feature_key.strip()
    if not raw:
        raise FeatureKeyError("featureKey 不能为空")
    segments = raw.split(SEPARATOR)
    if len(segments) != 4:
        raise FeatureKeyError(
            "featureKey 必须为「领域/系统/场景/功能点」四级，当前 "
            + str(len(segments))
            + " 段："
            + str(feature_key)
        )
    return of(segments[0], segments[1], segments[2], segments[3])


def join(domain: str, system: str, scene: str, feature: str) -> str:
    return of(domain, system, scene, feature).to_key()


def encode_for_url(feature_key: str) -> str:
    return quote_plus(parse(feature_key).to_key(), safe="")


def decode_from_url(encoded: str | None) -> str:
    decoded = unquote_plus("" if encoded is None else encoded)
    return parse(decoded).to_key()


def from_path_variable(feature_key: str | None) -> str:
    raw = "" if feature_key is None else feature_key.strip()
    if raw.startswith(SEPARATOR):
        raw = raw[1:]
    if "%" in raw:
        return decode_from_url(raw)
    return parse(raw).to_key()


def is_valid(feature_key: str) -> bool:
    try:
        parse(feature_key)
        return True
    except FeatureKeyError:
        return False


def java_string_hash(text: str) -> int:
    hash_value = 0
    for character in text:
        hash_value = ((hash_value * 31) + ord(character)) & 0xFFFFFFFF
    if hash_value >= 0x80000000:
        hash_value -= 0x100000000
    return hash_value


def java_hex(value: int) -> str:
    return format(value & 0xFFFFFFFF, "x")
